packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/CustomTabsConnectionHelper.kt

```diff
@@ -1,10 +1,98 @@
 package expo.modules.webbrowser
 
+import android.content.ComponentName
+import android.content.Context
 import android.net.Uri
-import expo.modules.core.interfaces.InternalModule
+import androidx.browser.customtabs.CustomTabsClient
+import androidx.browser.customtabs.CustomTabsServiceConnection
+import androidx.browser.customtabs.CustomTabsSession
 
-interface CustomTabsConnectionHelper : InternalModule {
-  fun warmUp(packageName: String)
-  fun mayInitWithUrl(packageName: String, uri: Uri)
-  fun coolDown(packageName: String): Boolean
+internal class CustomTabsConnectionHelper(
+  private val context: Context
+) : CustomTabsServiceConnection() {
+  private var currentPackageName: String? = null
+  private val clientActions = DeferredClientActionsQueue<CustomTabsClient>()
+  private val sessionActions = DeferredClientActionsQueue<CustomTabsSession?>()
+
+  // region lifecycle methods
+  fun destroy() = unbindService()
+  // endregion
+
+  // region Actual connection helper methods
+  fun warmUp(packageName: String) {
+    clientActions.executeOrQueueAction { client: CustomTabsClient -> client.warmup(0) }
+    ensureConnection(packageName)
+  }
+
+  fun mayInitWithUrl(packageName: String, uri: Uri) {
+    sessionActions.executeOrQueueAction { session: CustomTabsSession? ->
+      session?.mayLaunchUrl(uri, null, null)
+    }
+    ensureConnection(packageName)
+    ensureSession()
+  }
+
+  fun coolDown(packageName: String): Boolean {
+    if (packageName == currentPackageName) {
+      unbindService()
+      return true
+    }
+    return false
+  }
+  // endregion
+
+  // region CustomTabsServiceConnection implementation
+  override fun onBindingDied(componentName: ComponentName) {
+    if (componentName.packageName == currentPackageName) {
+      clearConnection()
+    }
+  }
+
+  override fun onCustomTabsServiceConnected(componentName: ComponentName, client: CustomTabsClient) {
+    if (componentName.packageName == currentPackageName) {
+      clientActions.setClient(client)
+    }
+  }
+
+  override fun onServiceDisconnected(componentName: ComponentName) {
+    if (componentName.packageName == currentPackageName) {
+      clearConnection()
+    }
+  }
+  // endregion
+
+  private fun ensureSession() {
+    if (sessionActions.hasClient()) {
+      return
+    }
+
+    clientActions.executeOrQueueAction { client: CustomTabsClient ->
+      sessionActions.setClient(client.newSession(null))
+    }
+  }
+
+  private fun ensureConnection(packageName: String) {
+    if (currentPackageName != null && currentPackageName != packageName) {
+      clearConnection()
+    }
+    if (!connectionStarted(packageName)) {
+      CustomTabsClient.bindCustomTabsService(context, packageName, this)
+      currentPackageName = packageName
+    }
+  }
+
+  private fun connectionStarted(packageName: String): Boolean {
```

Comment:
```suggestion
  private fun isConnectionStarted(packageName: String): Boolean {
```
It sounds better, doesn't it?